File: src/flake_analysis/api/schemas/explorer.py
"""Explorer schemas per backend design §1.2 + §1.3 + mosaic-viewer §3-§4."""
from __future__ import annotations
from typing import Optional
from pydantic import BaseModel, Field


class TileManifestEntry(BaseModel):
    image_id: int
    stem: str
    col: int = Field(ge=0)
    row: int = Field(ge=0)
    width_px: int = Field(gt=0)
    height_px: int = Field(gt=0)
    # LOD index (str) -> [w, h] pixel size of the cached thumbnail.
    lod_sizes: dict[str, list[int]]


class TileManifest(BaseModel):
    grid_w: int = Field(gt=0, le=60)  # Pinned decision #7: 60×60 cap
    grid_h: int = Field(gt=0, le=60)
    lod_sizes: dict[str, list[int]]
    signature: list[str]
    params_hash: str
    tiles: list[TileManifestEntry]


class ExplorerFlakeRow(BaseModel):
    flake_id: int
    image_id: int
    domains: int = Field(ge=0)
    groups: str
    distance: str
    clipped: str
    # 'pass' is a Python keyword — quote it on construction.
    # serialize_by_alias=True makes model_dump() emit "pass" (not "pass_").
    model_config = {"populate_by_name": True, "serialize_by_alias": True}
    pass_: bool = Field(alias="pass")


class ExplorerFlakesResponse(BaseModel):
    rows: list[ExplorerFlakeRow]
    total: int = Field(ge=0)


class ExplorerFlakeDetail(BaseModel):
    flake_id: int
    image_id: int
    domain_ids: list[int]
    cluster_names: list[str]
    bbox_xy: list[int]  # [x, y, w, h]
    mask_stats: dict[str, float]
    distance_px: Optional[float] = None
    isolation_px: Optional[float] = None


class NeighborFilterParams(BaseModel):
    size_min: Optional[int] = Field(default=None, ge=1)
    size_max: Optional[int] = Field(default=None, ge=1)
    isolation_min: Optional[float] = Field(default=None, ge=0.0)
    exclude_border_clipped: bool = False


class SaveExplorerStateParams(BaseModel):
    include_labels: list[str]
    exclude_labels: list[str]
    neighbor_filter: NeighborFilterParams
    selected_flake_ids: Optional[list[int]] = None


class SaveExplorerStateResult(BaseModel):
    state_path: str
    selected_count: Optional[int] = None

File: web/src/api/explorer.ts
// web/src/api/explorer.ts
import { ApiError } from '@/api/selector'

async function unwrap<T>(resp: Response): Promise<T> {
  if (!resp.ok) {
    let envelope: any = null
    try {
      envelope = await resp.json()
    } catch {
      throw new ApiError(resp.status, 'http_error', `HTTP ${resp.status}`, null)
    }
    const err = envelope?.error ?? {}
    throw new ApiError(
      resp.status,
      err.code ?? 'http_error',
      err.message ?? `HTTP ${resp.status}`,
      err.details ?? null,
      err.request_id
    )
  }
  return (await resp.json()) as T
}

export interface TileManifestEntryDto {
  image_id: number
  stem: string
  col: number
  row: number
  width_px: number
  height_px: number
  lod_sizes: Record<string, [number, number]>
}

export interface TileManifestDto {
  grid_w: number
  grid_h: number
  lod_sizes: Record<string, [number, number]>
  signature: string[]
  params_hash: string
  tiles: TileManifestEntryDto[]
}

export interface ExplorerFlakeRowDto {
  flake_id: number
  image_id: number
  domains: number
  groups: string
  distance: string
  clipped: string
  pass: boolean
}

export interface ExplorerFlakesResponseDto {
  rows: ExplorerFlakeRowDto[]
  total: number
}

export interface ExplorerFlakeDetailDto {
  flake_id: number
  image_id: number
  domain_ids: number[]
  cluster_names: string[]
  bbox_xy: number[]
  mask_stats: Record<string, number>
  distance_px: number | null
  isolation_px: number | null
}

export interface SaveExplorerStateBody {
  include_labels: string[]
  exclude_labels: string[]
  neighbor_filter: {
    size_min: number | null
    size_max: number | null
    isolation_min: number | null
    exclude_border_clipped: boolean
  }
  selected_flake_ids?: number[]
}

export interface SaveExplorerStateResultDto {
  state_path: string
  selected_count: number | null
}

export async function fetchTileManifest(projectId: string): Promise<TileManifestDto> {
  const resp = await fetch(
    `/api/v1/projects/${projectId}/explorer/tile_manifest`,
    { headers: { Accept: 'application/json' } }
  )
  return unwrap<TileManifestDto>(resp)
}

export async function fetchExplorerGrid(projectId: string): Promise<TileManifestDto> {
  const resp = await fetch(
    `/api/v1/projects/${projectId}/explorer/grid`,
    { headers: { Accept: 'application/json' } }
  )
  return unwrap<TileManifestDto>(resp)
}

export interface ExplorerFlakesQuery {
  include: string[]
  exclude: string[]
  sizeMin: number | null
  sizeMax: number | null
}

export async function fetchExplorerFlakes(
  projectId: string,
  q: ExplorerFlakesQuery
): Promise<ExplorerFlakesResponseDto> {
  const params = new URLSearchParams()
  if (q.include.length > 0) params.set('include', q.include.join(','))
  if (q.exclude.length > 0) params.set('exclude', q.exclude.join(','))
  if (q.sizeMin !== null) params.set('size_min', String(q.sizeMin))
  if (q.sizeMax !== null) params.set('size_max', String(q.sizeMax))
  const qs = params.toString()
  const url = `/api/v1/projects/${projectId}/explorer/flakes${qs ? `?${qs}` : ''}`
  const resp = await fetch(url, { headers: { Accept: 'application/json' } })
  return unwrap<ExplorerFlakesResponseDto>(resp)
}

export async function fetchExplorerFlakeDetail(
  projectId: string,
  flakeId: number
): Promise<ExplorerFlakeDetailDto> {
  const resp = await fetch(
    `/api/v1/projects/${projectId}/explorer/flake/${flakeId}`,
    { headers: { Accept: 'application/json' } }
  )
  return unwrap<ExplorerFlakeDetailDto>(resp)
}

export async function saveExplorerState(
  projectId: string,
  body: SaveExplorerStateBody
): Promise<SaveExplorerStateResultDto> {
  const resp = await fetch(
    `/api/v1/projects/${projectId}/run/explorer/save_state`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', Accept: 'application/json' },
      body: JSON.stringify(body),
    }
  )
  return unwrap<SaveExplorerStateResultDto>(resp)
}

export async function getExplorerState(projectId: string): Promise<any> {
  const resp = await fetch(
    `/api/v1/projects/${projectId}/run/explorer/state`,
    { headers: { Accept: 'application/json' } }
  )
  return unwrap<any>(resp)
}
